src/providers/postgres/qgspostgresprovider.cpp

```
@@ -3895,117 +3895,92 @@ QgsBox3D QgsPostgresProvider::extent3D() const
-              // estimated_extent() returns eastern bound of data (>-180) and
-              // 180 degrees.
-              if ( !ext.startsWith( QLatin1String( "-180 " ) ) && ext.contains( QLatin1String( ",180 " ) ) )
-              {
-                ext.clear();
-              }
-            }
-          }
-          else
-          {
-            // no features => ignore estimated extent
-            ext.clear();
-          }
-        }
-      }
-      else
-      {
-        QgsDebugMsgLevel( QStringLiteral( "no column statistics for %1.%2.%3" ).arg( mSchemaName, mTableName, mGeometryColumn ), 2 );
-      }
+
+      if ( result.PQresultStatus() != PGRES_TUPLES_OK ) break;
+      if ( result.PQntuples() != 1 ) break;
+      if ( result.PQgetisnull( 0, 0 ) ) break;
+
+      ext = result.PQgetvalue( 0, 0 );
     }
+    while ( 0 );
 
-    if ( ext.isEmpty() )
+  // Compute the extents, if estimation failed or was disabled
+  if ( ext.isEmpty() )
+    do
     {
       sql = QStringLiteral( "SELECT %1(%2%3) FROM %4%5" )
-            .arg( connectionRO()->majorVersion() < 2 ? "extent" : "ST_3DExtent",
+            .arg( vmaj < 2 ? "extent" : "ST_3DExtent",
                   quotedIdentifier( mBoundingBoxColumn ),
                   ( mSpatialColType == SctPcPatch || mSpatialColType == SctGeography ) ? "::geometry" : "",
                   mQuery,
                   filterWhereClause() );
 
       result = connectionRO()->LoggedPQexec( "QgsPostgresProvider", sql );
-      if ( result.PQresultStatus() != PGRES_TUPLES_OK )
-        connectionRO()->LoggedPQexecNR( "QgsPostgresProvider", QStringLiteral( "ROLLBACK" ) );
-      else if ( result.PQntuples() == 1 && !result.PQgetisnull( 0, 0 ) )
-        ext = result.PQgetvalue( 0, 0 );
+      if ( result.PQresultStatus() != PGRES_TUPLES_OK ) break;
+      if ( result.PQntuples() != 1 ) break;
+      if ( result.PQgetisnull( 0, 0 ) ) break;
+      ext = result.PQgetvalue( 0, 0 );
     }
+    while ( 0 );
```

Why are you using do/while? here and in the block above?

---

I find it more readable when there are a lot of reasons to skip a block.
More readable than a lot of nested conditions

---

I'm personally -0 to this coding style. I can see the logical benefit in it, but it's so rarely used that i find there's always difficulty in reading it and comprehending why something which runs once is using a loop statement...

---

Would you prefer goto statements ? Or some macro to make it more clear it's about to run ONCE ?

---

Extracting to a function, or even to a lambda function?

---

Ok I'll see about a function, maybe we could just throw an exception if estimation cannot be performed, trying different heuristics aimed at estimating

---

Extraction to functions done with aa5e200d254a6acc4627d568d765f6bb1e333b84